Pass STATS_PASSWORD and STATS_USERNAME in a secret

pecameron opened the pull request:
The router statistics page is protected by a username/password
login. The username and password are passed in the clear in environment
variables that are displayed in the clear in the dc and pod.

This change places the username and password into a secret and mentions
the secret in the environment variables in the dc and pod. The
information is in the clear in the running router pod (oc rsh <router> env)

https://trello.com/c/oMw9La1n/502-3-sccfsi-haproxy-statistics-page-password-is-stored-in-cleartext-routersecurity

pecameron commented:
@knobunc PTAL

knobunc commented on pkg/generate/app/env.go:
```diff
@@ -104,6 +104,35 @@ func (e Environment) List() []kapi.EnvVar {
+	e[name] = kapi.EnvVar{Name: name, Value: value}
+}
+
+func (e EnvVars) AddSecret(name, secret, key string) {
```
Doc this please (and Add)

pecameron replied:
Done.

imcsk8 commented:
Looks good so far

pecameron commented:
/retest

knobunc commented:
/lgtm

pecameron commented:
Added test for --stats-password= and --stats-username=
Added test for new app.env interfaces, Add(), AddSecret(), List()
@knobunc PTAL, you may want to get added to pkg/generate/OWNERS and pkg/oc/OWNERS

knobunc commented:
/retest

pecameron commented:
/retest

pecameron commented:
/retest

knobunc commented on test/cmd/router.sh:
```diff
@@ -42,11 +42,14 @@ os::cmd::expect_failure_and_text 'oc adm router --dry-run --host-network=false -
 os::cmd::expect_success_and_text 'oc adm router --dry-run --host-network=false --host-ports=false --strict-sni -o yaml' 'ROUTER_STRICT_SNI'
 
+# stats username and password
+s::cmd::expect_success_and_text 'oc adm router --dry-run --stats-password="xxx" --stats-user="statman" -o yaml' 'name: router-stats'
```
Typo: `os::cmd`...  The leading o is missing.

This is causing your test failures.

pecameron commented:
@knobunc thanks for spotting that.

knobunc commented:
/retest

knobunc commented:
`"error: router could not be created; service account "router" is not allowed to access the host network on nodes, grant access with oc adm policy add-scc-to-user hostnetwork -z router"`

knobunc commented:
`"error: secrets "router-stats" already exists"`

knobunc commented:
/approve

knobunc commented:
/assign @csrwng 

Cesar, can you look over (and hopefully, approve) the changes to the generate code.

csrwng commented:
/lgtm
/approve

knobunc commented:
@csrwng Thanks for the prompt review and approval!

pecameron commented:
@eparis @knobunc we are down to gce test. can this be merged, or is there more to do?

pecameron commented:
did a rebase

csrwng commented:
/lgtm

csrwng commented:
@pecameron it should automatically merge once all tests pass

csrwng commented:
@pecameron the gce deployment is failing because the router is failing to come up. It could have something to do with this change.

pecameron commented:
@csrwng I haven't gone through the GCE router bringup. So any thoughts are welcome.
The change creates a new secret that contains the STATS username and password. That's about it. By default oadm router creates a user name and password. Alternatively the admin can supply them. In either event they are placed in a secret and referenced in the env vars. This works on bare metal and in the other testing, so it may be something that GCE does differently.

smarterclayton commented:
If you're making a change to how we deploy that is not backwards compatible with the old images, then you'll be unable to merge.

smarterclayton commented:
It looks like the failure is related to install, which means you aren't compatible with old ansible:

```
TASK [openshift_hosted : Poll for OpenShift router deployment success] *********
Wednesday 27 September 2017  16:20:23 +0000 (0:00:00.721)       0:27:03.349 *** 
FAILED - RETRYING: Poll for OpenShift router deployment success (40 retries left).
FAILED - RETRYING: Poll for OpenShift router deployment success (39 retries left).
failed: [ci-prtest-5a37c28-8744-ig-m-0fdg] (item=[{u'name': u'router', u'certificate': {u'certfile': u'/etc/origin/master/openshift-router.crt', u'keyfile': u'/etc/origin/master/openshift-router.key', u'cafile': u'/etc/origin/master/ca.crt'}, u'replicas': u'1', u'serviceaccount': u'router', u'namespace': u'default', u'stats_port': 1936, u'edits': [{u'action': u'put', u'key': u'spec.strategy.rollingParams.intervalSeconds', u'value': 1}, {u'action': u'put', u'key': u'spec.strategy.rollingParams.updatePeriodSeconds', u'value': 1}, {u'action': u'put', u'key': u'spec.strategy.activeDeadlineSeconds', u'value': 21600}], u'images': u'openshift/origin-${component}:${version}', u'selector': u'role=infra', u'ports': [u'80:80', u'443:443']}, {'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2017-09-27 16:20:23.037950', '_ansible_no_log': False, u'stdout': u'1', u'changed': True, u'start': u'2017-09-27 16:20:22.675143', u'delta': u'0:00:00.362807', u'cmd': [u'oc', u'get', u'deploymentconfig', u'router', u'--namespace', u'default', u'--config', u'/etc/origin/master/admin.kubeconfig', u'-o', u'jsonpath={ .status.latestVersion }'], 'item': {u'name': u'router', u'certificate': {u'keyfile': u'/etc/origin/master/openshift-router.key', u'certfile': u'/etc/origin/master/openshift-router.crt', u'cafile': u'/etc/origin/master/ca.crt'}, u'replicas': u'1', u'serviceaccount': u'router', u'namespace': u'default', u'selector': u'role=infra', u'edits': [{u'action': u'put', u'value': 1, u'key': u'spec.strategy.rollingParams.intervalSeconds'}, {u'action': u'put', u'value': 1, u'key': u'spec.strategy.rollingParams.updatePeriodSeconds'}, {u'action': u'put', u'value': 21600, u'key': u'spec.strategy.activeDeadlineSeconds'}], u'images': u'openshift/origin-${component}:${version}', u'stats_port': 1936, u'ports': [u'80:80', u'443:443']}, u'rc': 0, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u"oc get deploymentconfig router --namespace default --config /etc/origin/master/admin.kubeconfig -o jsonpath='{ .status.latestVersion }'", u'removes': None, u'creates': None, u'chdir': None}}, 'stdout_lines': [u'1'], u'stderr': u''}]) => {"attempts": 3, "changed": true, "cmd": ["oc", "get", "replicationcontroller", "router-1", "--namespace", "default", "--config", "/etc/origin/master/admin.kubeconfig", "-o", "jsonpath={ .metadata.annotations.openshift\\.io/deployment\\.phase }"], "delta": "0:00:00.226851", "end": "2017-09-27 16:20:54.768652", "failed": true, "failed_when_result": true, "item": [{"certificate": {"cafile": "/etc/origin/master/ca.crt", "certfile": "/etc/origin/master/openshift-router.crt", "keyfile": "/etc/origin/master/openshift-router.key"}, "edits": [{"action": "put", "key": "spec.strategy.rollingParams.intervalSeconds", "value": 1}, {"action": "put", "key": "spec.strategy.rollingParams.updatePeriodSeconds", "value": 1}, {"action": "put", "key": "spec.strategy.activeDeadlineSeconds", "value": 21600}], "images": "openshift/origin-${component}:${version}", "name": "router", "namespace": "default", "ports": ["80:80", "443:443"], "replicas": "1", "selector": "role=infra", "serviceaccount": "router", "stats_port": 1936}, {"_ansible_item_result": true, "_ansible_no_log": false, "_ansible_parsed": true, "changed": true, "cmd": ["oc", "get", "deploymentconfig", "router", "--namespace", "default", "--config", "/etc/origin/master/admin.kubeconfig", "-o", "jsonpath={ .status.latestVersion }"], "delta": "0:00:00.362807", "end": "2017-09-27 16:20:23.037950", "invocation": {"module_args": {"_raw_params": "oc get deploymentconfig router --namespace default --config /etc/origin/master/admin.kubeconfig -o jsonpath='{ .status.latestVersion }'", "_uses_shell": false, "chdir": null, "creates": null, "executable": null, "removes": null, "warn": true}}, "item": {"certificate": {"cafile": "/etc/origin/master/ca.crt", "certfile": "/etc/origin/master/openshift-router.crt", "keyfile": "/etc/origin/master/openshift-router.key"}, "edits": [{"action": "put", "key": "spec.strategy.rollingParams.intervalSeconds", "value": 1}, {"action": "put", "key": "spec.strategy.rollingParams.updatePeriodSeconds", "value": 1}, {"action": "put", "key": "spec.strategy.activeDeadlineSeconds", "value": 21600}], "images": "openshift/origin-${component}:${version}", "name": "router", "namespace": "default", "ports": ["80:80", "443:443"], "replicas": "1", "selector": "role=infra", "serviceaccount": "router", "stats_port": 1936}, "rc": 0, "start": "2017-09-27 16:20:22.675143", "stderr": "", "stderr_lines": [], "stdout": "1", "stdout_lines": ["1"]}], "rc": 0, "start": "2017-09-27 16:20:54.541801", "stderr": "", "stderr_lines": [], "stdout": "Failed", "stdout_lines": ["Failed"]}
```

pecameron commented:
@smarterclayton Thanks for taking a look at this. 
The changes set up a new secret for the stats uname and password. By default nothing has to be done, everything is handled internally. Could you take a look at the change to see if it is OK?

pecameron commented:
@smarterclayton I took a look at the failure. I don't understand what caused the failure. Do you have any idea?

pecameron commented:
@knobunc cloned/built/installed pr/16389
created router - it works
# oc adm router router-phil1
info: password for stats user admin has been set to 8jYh8Pxznc
--> Creating router router-phil1 ...
    secret "router-phil1-stats" created
    warning: serviceaccounts "router" already exists
    clusterrolebinding "router-router-phil1-role" created
    deploymentconfig "router-phil1" created
    service "router-phil1" created
--> Success
What is different in gce?

pecameron commented:
@smarterclayton @stevekuznetsov @knobunc - I am stuck. I need help with this. I have very little experience with gce or how this test works.  If you folks are not the right people could you point me to someone that could possibly help. Thanks!

smarterclayton commented:
Gce installs with Ansible.  If your change has altered something
Ansible depends on, you'll need to get advice from the Ansible team on
how your patch may have broken application installs.  The
install_update job doesn't exercise the same depth of setup as gce, so
we run both to ensure nothing breaks.

pecameron commented:
@stevekuznetsov could you point me to the ansible that is having the problem?  Is there a doc with how to work with this? I have not worked in this space before so any help getting into it is appreciated. Thanks!

stevekuznetsov commented:
@pecameron looks like you are failing [here](https://github.com/openshift/openshift-ansible/blob/acb78312d18c27f607ad71898fcc6bacab4b5cb6/roles/openshift_hosted/tasks/router/router.yml#L112) -- I would reach out to @sdodson  and @abutcher to help with this

pecameron commented:
@sdodson @abutcher Could you help me find the cause of this gce test failure? I am at at a loss. The only thing that is really new is the creation of a new opaque secret, <router-name>-stats that holds the stats username and password. I am wondering if that might be causing this failure? Thanks!
CC: @knobunc @stevekuznetsov

stevekuznetsov commented:
/lgtm cancel

abutcher commented:
Talked with @kwoodson about this and I think we'll need to update [oc_adm_router.py](https://github.com/openshift/openshift-ansible/blob/master/roles/lib_openshift/library/oc_adm_router.py) within openshift-ansible to handle the new secret. The `router-stats` secret wasn't created when I tested locally with the packages from the failing run.

```
# oc get secret router-stats
Error from server (NotFound): secrets "router-stats" not found
```

pecameron commented:
@knobunc @abutcher  The gce fail will take a change in ansible. What do you want to do with this PR? It is a 3.7 item.

smarterclayton commented:
I'm not totally happy with a new secret to add the ever growing list.  I
don't think this is a required 3.7 item since there is a workaround that
someone can follow.

On Thu, Sep 28, 2017 at 5:26 PM, Phil Cameron <notifications@github.com>
wrote:

> @knobunc <https://github.com/knobunc> @abutcher
> <https://github.com/abutcher> The gce fail will take a change in ansible.
> What do you want to do with this PR? It is a 3.7 item.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/16389#issuecomment-332968576>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3e0m5uyQmJxuTqPHyuj4Eulo3-Fks5snA70gaJpZM4PZd62>
> .
>
